From 056f788ea3a9ba1b45e17bcacea91a38c1ed8d73 Mon Sep 17 00:00:00 2001
From: Christian Beer <christian.beer@aei.mpg.de>
Date: Mon, 18 Jul 2016 16:01:26 +0200
Subject: [PATCH] Tools: fix segfault in crypt_prog

The API changes in e965ea2 introduced a segfault when converting a private BOINC key into OpenSSL format. Also moved from the deprecated (since at least 1.0) RSA_generate_key() function to RSA_generate_key_ex().
---
 lib/crypt_prog.cpp | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/lib/crypt_prog.cpp b/lib/crypt_prog.cpp
index 3bc2d53..88d9f2d 100644
--- a/lib/crypt_prog.cpp
+++ b/lib/crypt_prog.cpp
@@ -126,7 +126,7 @@ int main(int argc, char** argv) {
     FILE *f, *fpriv, *fpub;
     char cbuf[256];
 #ifdef HAVE_OPAQUE_RSA_DSA_DH
-    RSA *rsa_key;
+    RSA *rsa_key = RSA_new();
 #else
     RSA rsa_key;
 #endif
@@ -136,6 +136,7 @@ int main(int argc, char** argv) {
     char *certpath;
     bool b2o=false; // boinc key to openssl key ?
     bool kpriv=false; // private key ?
+    BIGNUM *e;
 
     if (argc == 1) {
         usage();
@@ -150,7 +151,16 @@ int main(int argc, char** argv) {
         n = atoi(argv[2]);
 
         srand(random_int());
-        RSA* rp = RSA_generate_key(n,  65537, 0, 0);
+        e = BN_new();
+        retval = BN_set_word(e, (unsigned long)65537);
+        if (retval != 1) {
+            die("BN_set_word");
+        }
+        RSA *rp = RSA_new();
+        retval = RSA_generate_key_ex(rp, n, e, NULL);
+        if (retval != 1) {
+            die("RSA_generate_key_ex");
+        }
         openssl_to_keys(rp, n, private_key, public_key);
         fpriv = fopen(argv[3], "w");
         if (!fpriv) die("fopen");
